libs/libc/semaphore/sem_open.c

```diff
@@ -112,7 +112,7 @@ FAR sem_t *sem_open(FAR const char *name, int oflags, ...)
   /* Let nxsem_open() do the work */
 
   sem = nxsem_open(name, oflags, mode, value);
-  if (sem < 0)
+  if ((intptr_t)sem < 0 && (intptr_t)sem > -__ELASTERROR)
```

Comment:
Could you please explain more on the `-__ELASTERROR` comparison? PR description is now misleading as states only about type conversion

Reply:
I think the problem is of my creation, returning a negative pointer and doing comparisons against it was maybe not the best of ideas.

In u32 arithmetics, a pointer ptr=0x80000000 is just fine, but comparing if ((intptr_t)ptr < 0) returns true in this case. Doing the -__ELASTERROR removes some of the range where the calling logic erroneously thinks the pointer is a negative number, but not all of it. Also, a pointer between [0x80000000, 0x80000000+__ELASTERROR] is also a valid pointer.

I think this needs to be fixed properly, by fixing nx_semopen().

Reply:
Yes, agreed.

Reply:
Or `sem_t** sem` parameter as an alternative and return error code or zero, but I prefer to add `int *error` parameter.